### DISCUSSION TITLE
Video Writer Node Video Compression Quality Tuning

### DISCUSSION THREAD

**Discussion by saeyokoyama** [bonsai-rx/bonsai#1778] [Q&A]
I'm writing to ask how to control video compression bitrate on bonsai, if that is not possible, are there any suggestions on codec codes that compress less than DIVX and FFV1 but more than H.265? 

I have two FLIR cameras to record animal behavior in addition to an internal camera inside an Neurophotometric machine which captures photometry data. All three cameras are controlled by Bonsai. The NPM camera captures at 20Hz for 250ms every 5 mins (takes only 5 frames per 5 min). Behavior Camera 1 captures at 20Hz for 5 seconds every 5 min (100 frames per 5 min). Behavior Camera 2 captures at 40Hz and is manually controlled (can start acquiring On and Off by a key click). I want Behavior Camera 2 to capture for 1-2 hours. 

Currently, Video Writer's FourCC  is set to 0.

All of the data from the cameras are saved on one computer. When the NPM camera and Camera 1 are on, I have no problems with data acquisition. The moment I turn on Camera 2, when NPM camera and Camera 1 are acquiring, Disk 1 (Crucial P3 2 PCIe NVMe SSD) reaches 100% and the average response time becomes increasingly slow. The memory on this computer seems to be functioning just fine. I am hoping to run SLEAP on this video data to extract behavior. 

Therefore, I would like to compress my Camera 2 acquisition minimally to solve this disk issue to have  high quality, almost uncompressed videos. I've tried a couple of other codec codes (S, 0, H.265, DIB) and those still cause Disk1 to saturate. Using  FFV1, the .avi video did not play, and with DIVX and FFV1 produced significantly smaller videos but the video quality is lower than I'd like it to be. I already compress my videos after capturing using FFmpeg but again my main issue is that the disk saturates during capture and crashes bonsai. 

The node where the FourCC code is indicated: 
![Screenshot 2024-05-06 130109](https://github.com/bonsai-rx/bonsai/assets/122409575/ade9f2b5-8b0e-42b4-a599-694ee461b471)

A list of of codes I've tried, and their outcomes:
![image](https://github.com/bonsai-rx/bonsai/assets/122409575/c41bac6a-5110-4a42-9ac9-0bbe939404d0)

Example of what happens when the disk becomes saturated: 
![image](https://github.com/bonsai-rx/bonsai/assets/122409575/ddd78135-8f3d-4653-8ae3-eea309593623)

**Comment by bruno-f-cruz** [comment]
I would suggest using the FFMPEG `VideoWriter` node instead and set all these parameters in the `OutputParameters` property.

**Reply by saeyokoyama** [reply]
Thank you for the reply. I didn't realize this tool existed. Is the "OutputArguments" where I input my ffmpeg settings to compress the video? I'm unclear on how to use this node.

Is this: 
![image](https://github.com/bonsai-rx/bonsai/assets/122409575/045e75b7-e342-4112-84d2-2cddfe5fe6a4)

Where I input this? 
ffmpeg -i [video].avi -filter:v hqdn3d=4:4:3:3 -pix_fmt yuv420p10le -c:v libx265 -crf 16  [compressed video].avi

**Reply by bruno-f-cruz** [reply]
`filename` =  [compressed video].avi
`OutputArguments` = `-filter:v hqdn3d=4:4:3:3 -pix_fmt yuv420p10le -c:v libx265 -crf 16`

**Reply by saeyokoyama** [reply]
Thank you! 

I am now running into this error: 
![image](https://github.com/bonsai-rx/bonsai/assets/122409575/31ff2b01-8e15-44e0-b897-7b1cb44f992a)

**Reply by bruno-f-cruz** [reply]
You need to have ffmpeg available to the application path. Either make it globally available via Environment Variables or add the executable to the bonsai Extensions folder.

**Reply by saeyokoyama** [reply]
The way I was able to add ffmpeg Video Writer onto my workflow was by saving the workflow file into the  Bonsai.Example.FFmpegWriter. However, when I add all of the components within Bonsai.Example.FFmpegWriter to the extensions file associated with Bonsai, it does not appear within the toolbox. 

![image](https://github.com/bonsai-rx/bonsai/assets/122409575/a5e00ede-f52a-4f5e-b3d7-ace3ad2000c1)

![image](https://github.com/bonsai-rx/bonsai/assets/122409575/09ab4974-ecbc-41a6-9c56-35856d0913b0)

**Reply by bruno-f-cruz** [reply]
Not sure I understand how you are organizing the project. You don't need the ffmpeg example to use the videowriter operator.
Maybe reading the documentation helps clearing up some of the confusion? https://bonsai-rx.org/docs/articles/scripting-extensions.html

**Reply by saeyokoyama** [reply]
I apologize for the confusion, I'm super new to bonsai and I've never added a extension before. I didn't know there were extensions until today. Out of the files within "Bonsai.Examples.FFmpegWriter" which one is the operator? 
More simply, which one of these: 
![image](https://github.com/bonsai-rx/bonsai/assets/122409575/3f423853-4335-4000-8898-ef53d17d49a2)

Am I supposed to add into here: 
![image](https://github.com/bonsai-rx/bonsai/assets/122409575/06a66bf3-5ba9-4f89-9d22-02acb080eba5)

and by project, i assume you mean my bonsai workblow and layout file? They are currently in a separate directory under "documents" in my bootdrive.

**Reply by bruno-f-cruz** [reply]
You just need to install the FFMPEG package from the package manager (see instructions here https://bonsai-rx.org/docs/articles/packages.html) and the `VideoWriter` node will appear in the toolbox.

**Reply by saeyokoyama** [reply]
Thank you for taking the time to walk through the steps! I was able to get this to work and is exactly what I was looking for!
